Add modules overview (#135)
bcnmy/docs: docs/tutorials/customValidationModule.md

# Custom Validation Module

In this tutorial we will look at how to create your own validation module. 
Any custom validation module needs to extend abstract contract [BaseAuthorizationModule](https://github.com/bcnmy/scw-contracts/blob/main/contracts/smart-account/modules/BaseAuthorizationModule.sol) - which extends IAuthorizationModule and ISignatureValidator interface.

> The implementation of this interface enables a module to receive userOp data and provide validation results back to the SmartAccount. 

Reviewer: `VALIDATION_SUCCESS` and `VALIDATION_FAILED`:

Validation Status Constants: These constants are imported from the `AuthorizationModulesConstants` contract. `VALIDATION_SUCCESS` indicates successful validation, while `VALIDATION_FAILED` signifies validation failure.

> The implementation of this interface enables a module to receive userOp data and provide validation results back to the SmartAccount. 

Author: I will add the above information on the different interfaces in the overview page itself, given in the tutorial we mainly want to focus on the code.

Lets take a look at the step by step guide to create a custom Validation Module.

> **1. Create the custom module class**

Reviewer: not `class`, should be something like `custom validation contract`

```typescript
import {BaseAuthorizationModule} from "./BaseAuthorizationModule.sol";

contract MyCustomValidationModule is BaseAuthorizationModule {
    string public constant NAME = "My Custom Validation Module";
    string public constant VERSION = "0.0.1";

    // Function signature of this method will be used as moduleSetupData in Account Factory
    function initForSmartAccount() {
      
    }
}
```

> As per the code
> - First we import the BaseAuthorizationModule and initialise two constants name and version
> - Then we declare a method initForSmartAccount which gets used to initialise this module for the smart account. SA calls this function with msg.sender as smart account address. It updates the relevant storage for the msg sender, It could be ownership information as shown in ECDSAOwnsership Module.

Reviewer 2: ```suggestion
As per the code
- First we import the BaseAuthorizationModule and initialize two constants name and version
- Then we declare a method initForSmartAccount which gets used to initialize this module for the smart account. SA calls this function with msg.sender as the smart account address. It updates the relevant storage for the msg sender, It could be ownership information as shown in ECDSAOwnsership Module.
```

**2. Implement validateUserOp method**
```typescript
/// @inheritdoc IAuthorizationModule
function validateUserOp(
    UserOperation calldata userOp,
    bytes32 userOpHash
) external view virtual override returns (uint256) {

    (bytes memory moduleSignature, ) = abi.decode(
        userOp.signature,
        (bytes, address)
    );
    // implement _verifySignature as per the module functionality
    if (_verifySignature(userOpHash, moduleSignature, userOp.sender)) {
        return VALIDATION_SUCCESS;
    }
    return SIG_VALIDATION_FAILED;
}
```

> As per the code
> - This functions requires the userOp and the userOp Hash as the input parameters where userOp signature being the ABI-encoded signature and validation module address. we extract the module signature and call the verify signature method which we will implement next.
> - `_verifySignature` method expects moduleSignature which is a signature that should be processed by a module and made according to the requirements specified by the module that is expected to be processing it.

Reviewer 2: ```suggestion
As per the code
- This function requires the userOp and the userOp Hash as the input parameters where userOp signature is the ABI-encoded signature and validation module address. we extract the module signature and call the verify signature method which we will implement next.
- `_verifySignature` method expects moduleSignature which is a signature that should be processed by a module and made according to the requirements specified by the module that is expected to be processing it.
```

        }
        return bytes4(0xffffffff);
    }
```
As per the code
- `isValidSignatureUnsafe`: this method validates an EIP-1271 signature but expects the data Hash to already include smart account address information.

**5. Implement _verifySignature method**
```typescript
/**
* @param dataHash Hash of the data to be validated.
* @param signature Signature to be validated.
* @param smartAccount expected signer Smart Account address.
* @return true if signature is valid, false otherwise.
*/
function _verifySignature(
    bytes32 dataHash,
    bytes memory signature,
    address smartAccount
) internal view returns (bool) {

    // verification according to signature scheme of this module
}
```
As per the code `_verifySignature` method contains the custom logic for this module, which developer can implement based on their requirements.

Expand the code below to see the entire code:
<details>

```javascript

import {BaseAuthorizationModule} from "./BaseAuthorizationModule.sol";

contract MyCustomValidationModule is BaseAuthorizationModule {
    string public constant NAME = "My Custom Validation Module";
    string public constant VERSION = "0.0.1";

    /**

>         Smart contract account calls this function with msg.sender as smart account address
>         It updates the relevant storage for the msg sender, It could be ownership information as mentioned in ECDSAOwnsership Module.
>         Function signature of this method will be used as moduleSetupData in Account Factory

Reviewer 2: ```suggestion
        Smart contract account calls this function with msg.sender as smart account address
        It updates the relevant storage for the msg sender, It could be ownership information as mentioned in ECDSAOwnsership Module.
        The function signature of this method will be used as moduleSetupData in Account Factory
```

                keccak256(
                    abi.encodePacked(
                        "\x19Ethereum Signed Message:\n52",
                        dataHash,
                        smartAccount
                    )
                ),
                moduleSignature,
                smartAccount
            )
        ) {
            return EIP1271_MAGIC_VALUE;
        }
        return bytes4(0xffffffff);
    }

    /// @inheritdoc ISignatureValidator
    function isValidSignatureUnsafe(
        bytes32 dataHash,
        bytes memory moduleSignature
    ) public view virtual returns (bytes4) {
        return
            isValidSignatureForAddressUnsafe(
                dataHash,
                moduleSignature,
                msg.sender
            );
    }

    /// @inheritdoc IEcdsaOwnershipRegistryModule
    function isValidSignatureForAddressUnsafe(
        bytes32 dataHash,
        bytes memory moduleSignature,
        address smartAccount
    ) public view virtual returns (bytes4) {
        if (_verifySignature(dataHash, moduleSignature, smartAccount)) {
            return EIP1271_MAGIC_VALUE;
        }
        return bytes4(0xffffffff);
    }
    
}
```
</details>

> Note: We could also create module's own interface IMyCustomValidationModule to move storage, events and errors there and then make the above module extend this interface. 

Reviewer 2: ```suggestion
Note: We could also create a module's own interface IMyCustomValidationModule to move storage, events and errors there and then make the above module extend this interface. 
```

> # Custom Validation Module
> 
> In this tutorial we will look at how to create your own validation module.
> 
> As mentioned in the [overview](./../modules#how-to-create-a-custom-validation-module), any custom validation module needs to extend the abstract contract [BaseAuthorizationModule](https://github.com/bcnmy/scw-contracts/blob/main/contracts/smart-account/modules/BaseAuthorizationModule.sol). This contract is crucial as it connects your module to essential authorization functions.
> The following implementation enables a module to receive userOp data and provide validation results back to the SmartAccount.
> 
> Lets take a look at the step by step guide to create a custom Validation Module.
> 
> **1. Create the custom validation contract**

Reviewer 2: ```suggestion
# Custom Validation Module

In this tutorial, we will look at how to create your own validation module.

As mentioned in the [overview](./../modules#how-to-create-a-custom-validation-module), any custom validation module needs to extend the abstract contract [BaseAuthorizationModule](https://github.com/bcnmy/scw-contracts/blob/main/contracts/smart-account/modules/BaseAuthorizationModule.sol). This contract is crucial as it connects your module to essential authorization functions.
The following implementation enables a module to receive userOp data and provide validation results back to the SmartAccount.

Let's take a look at the step-by-step guide to create a custom Validation Module.

**1. Create the custom validation contract**
```

```typescript
import {BaseAuthorizationModule} from "./BaseAuthorizationModule.sol";

contract MyCustomValidationModule is BaseAuthorizationModule {
    string public constant NAME = "My Custom Validation Module";
    string public constant VERSION = "0.0.1";

    // Function signature of this method will be used as moduleSetupData in Account Factory
    function initForSmartAccount() {
      
    }
}
```

> As per the code
> 
> - First we import the BaseAuthorizationModule and initialize two constants name and version
> - Then we declare a method initForSmartAccount which gets used to initialize this module for the smart account. SA calls this function with msg.sender as the smart account address. It updates the relevant storage for the msg sender, It could be ownership information as shown in ECDSAOwnsership Module.
> 
> **2. Implement validateUserOp method**

Reviewer 2: ```suggestion
As per the code

- First, we import the BaseAuthorizationModule and initialize two constants name and version
- Then we declare a method initForSmartAccount which gets used to initialize this module for the smart account. SA calls this function with msg.sender as the smart account address. It updates the relevant storage for the msg sender, It could be ownership information as shown in ECDSAOwnsership Module.

**2. Implement validateUserOp method**
```
